### PR TITLE
Feature/basic features

### PR DESCRIPTION
**Pull request by bx44**
This pull request includes a basic implementation of different enumerable methods, as requested in the specifications of the project. It also includes the llinter configuration files and an informative Readme.md file as well.

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice progress :+1: 
This is your first attempt at this project and I am happy with the progress you have made so far. Let's work on some suggestions so we can merge and proceed.
Check my comments for suggestions.

### Martin Afani

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      my_each do |x|
+        return false unless x
+      end
+      return true
+    end
+    my_each do |x|
+      return false unless yield(x)
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9)
array.my_all?(Integer) == array.all?(Integer)
```
Should return `true` if `my_all?` works as expected.
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
Should return `true` if `my_all?` works as expected.
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3)
```
Should return `true` if `my_all?` works as expected.

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      my_each do |x|
+        return true if x
+      end
+      return false
+    end
+    my_each do |x|
+      return true if yield(x)
+    end
+    false
+  end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
array.my_any?(Integer) == array.any?(Integer)
```
Should return `true`
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/)
```
Should return `false`
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat')
```
Should return `false`

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      my_each do |x|
+        return false if x
+      end
+      return true
+    end
+    my_each do |x|
+      return false if yield(x)
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array.my_none?(String)
```
Should return`true`
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
words.my_none?(/z/) == words.none?(/z/)
```
Should return `true`
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words.my_none?(5) == words.none?(5)
```
Should also return `true`

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    new_array
+  end
+
+  def my_inject(initial = nil)
+    initial ||= self[0]
+    my_each do |x|
+      initial = yield(initial, x)
+    end
+    initial
+  end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array.my_inject(:+) == array.inject(:+)
```

Comment (bx44) [reply enumerable_methods.rb]: Thanks a lot for the review, I believe everything is fine now.

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress @bx44 :ok_hand: 
You have done a splendid job with the suggestions I left you. I only have one today. Implement it, and resubmit. Check my comments for suggestions

### Martin Afani

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      my_each do |x|
+        initial = x.send(sym, initial)
+      end
+      return initial
+    end
+    my_each do |x|
+      initial = yield(initial, x)
+    end
+    initial
+  end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] my_inject combines all elements of enum by applying a binary operation, specified by a block

Comment (bx44) [reply enumerable_methods.rb]: I believe it's fixed now

**Review by stratospherique — APPROVED** [review]
### Great work applying the suggested changes :clap: Congrats, the project is approved :heavy_check_mark: You may merge :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**

**Pull request merged**